### PR TITLE
Added caching

### PR DESCRIPTION
**Pull request by jchannon**
This PR caches the results to `MakeGenericType` to improve perf.

Thought I'd send this up for you to look at before I do similar work for `GetNotificationHandlers` 

cc // @khellang

### CODE REVIEW

File: src/MediatR/Mediator.cs [hunk 1]
```diff
@@ -17,6 +17,9 @@ public class Mediator : IMediator
 
         private readonly MultiInstanceFactory _multiInstanceFactory;
 
+        private readonly Dictionary<Type, Type> genericHandlerCache;
+        private readonly Dictionary<Type, Type> wrapperHandlerCache;
```
Comment (khellang) [inline src/MediatR/Mediator.cs]: `ConcurrentDictionary`?

Comment (jchannon) [inline src/MediatR/Mediator.cs]: Not available in PCL

**Comment by jchannon** [comment]
I also wonder whether we should cache the result of `Activator.CreateInstance`

File: src/MediatR/Mediator.cs (hunk 1 above)
Comment (jbogard) [inline src/MediatR/Mediator.cs]: Is it available in some profiles?

Comment (khellang) [inline src/MediatR/Mediator.cs]: @jbogard What's the reason for using Profile259? Would it make sense to switch to `dotnet5.1`, which essentially is Profile259, but will work for future platforms as well.

Comment (jchannon) [inline src/MediatR/Mediator.cs]: I have no clue about profiles/PCLs so I'll have to let you decide what's best

Comment (jbogard) [inline src/MediatR/Mediator.cs]: @khellang Just historical. I'm moving to .NET platform standard eventually, the question is whether to make a pitstop at `dotnet5.1` or not (I've had problems with it). It probably wouldn't hurt just to jump to it.

**Comment by jchannon** [comment]
Added caching for `publish` and `Activator.CreateInstance`

**Comment by jbogard** [comment]
I merged all the xproj stuff, can you merge/rebase?

**Comment by jchannon** [comment]
Will do. Any changes required? @khellang didn't like this the last time he
looked at it

**Comment by jbogard** [comment]
You'll likely be able to use the concurrent dictionary stuff now.

**Comment by khellang** [comment]
Yeah, that instance cache is bad. It effectively makes all handlers singletons, regardless of container lifetime scope. It's also horrible to cache by hash code. What about compiling new-expressions and caching those lambdas instead of calling `Activator.CreateInstance`?

**Comment by jbogard** [comment]
Also, one thing worth including are benchmarks about how much faster it is. Nearly all of my MediatR calls hit a DB or web service, so 1 vs 10 microseconds of reflection doesn't matter if I'm waiting 10-2000ms for an external call to finish.

**Comment by jchannon** [comment]
Could use this? https://github.com/jchannon/MediatRTest/

Shall we implement @khellang idea?

**Comment by khellang** [comment]
I just did a quick measurement of `Activator.CreateInstance` and compiled expressions on Windows. For default ctors, it doesn't matter, but for parameterized ctors, `Activator.CreateInstance` was about 46x slower. I ran the benchmark for 1.000.000 iterations and it completed in 774ms. At that point, I think we're talking about micro optimizations. 774ms on 1.000.000 requests is nothing.

**Comment by khellang** [comment]
BTW, @jbogard, have you seen https://github.com/dotnet/roslyn/pull/6644? When/if that lands, you'd be able to bypass reflection completely. You can just add a generic parameter:

```csharp
TResponse Send<TRequest, TResponse>(TRequest request) where TRequest : IRequest<TResponse>;
```

If you did this today, you'd end up with

```csharp
mediator.Send<Request, Response>(request);
```

Buf after the PR lands, you'd be able to infer all the type arguments

```csharp
mediator.Send(request);
```

**Comment by jbogard** [comment]
@khellang well crap let's just wait for that then

**Comment by jchannon** [comment]
That roslyn issue has been closed

**Comment by khellang** [comment]
It's still at https://github.com/dotnet/roslyn/pull/7850

**Comment by jchannon** [comment]
Hi, I've come back to this and tested without instance caching but with type caching and the results are

| Runtime  | Results  |   
|---|---|---|---|---|
| Mono 3.12.1 (with type caching)  | 63ms  |  
| Mono 4.2.2 (with type caching)  | 44ms  |   
| Mono 3.12.1 (nuget)  | 98ms  |   
| Mono 4.2.2 (nuget)  | 107ms  |   

The tests is here - https://github.com/jchannon/MediatRTest/

Would you consider merging this without the instance cache?

**Comment by khellang** [comment]
Compared to what without the caching?

**Comment by jchannon** [comment]
Sorry, yes. Tests with my PR and without using the nuget package

**Comment by Yantrio** [comment]
https://github.com/jchannon/MediatRTest/blob/master/Program.cs#L19

**Comment by jchannon** [comment]
Updated table with use of concurrent dictionary

**Comment by jchannon** [comment]
Have spoken to Mono team and I gave them a repro for ConcurrentDictionary and there seems to be a performance problem with it on 4.2.2 hence why we see a -32% slow down. I think its fixed in latest master and it appears to be quicker than 3.12.1.

However, hope you think the concurrentdictionary caching is useful and would love to see this merged

**Pull request merged**